compiler/plugins/target/AMD-AIE/aie/AIEDmaToNpu.cpp

```diff
@@ -24,16 +27,171 @@ using namespace xilinx::AIEX;
 #include "aie/Dialect/AIEX/Transforms/AIEXPasses.h.inc"
 #undef GEN_PASS_DEF_AIEDMATONPU
 
-namespace {
+#define TXN_OPC_WRITE 0x0
+#define TXN_OPC_BLOCKWRITE 0x1
+#define TXN_OPC_TCT 0x80
+#define TXN_OPC_DDR_PATCH 0x81
+
+// Example:
+// - instructions = {3,4,5}
+// - tailSize = 2
+// instructions becomes {3,4,5,0,0} and
+// a mutable reference to the tail {0,0} is returned.
+llvm::MutableArrayRef<uint32_t> reserveAndGetTail(
+    std::vector<uint32_t> &instructions, uint64_t tailSize) {
+  auto oldSize = instructions.size();
+  auto newSize = oldSize + tailSize;
+  instructions.resize(newSize, 0);
+  return llvm::MutableArrayRef<uint32_t>(instructions.data() + oldSize,
+                                         tailSize);
+}
+
+void appendSync(std::vector<uint32_t> &instructions, NpuSyncOp op) {
+  auto words = reserveAndGetTail(instructions, 4);
+
+  // XAIE_IO_CUSTOM_OP_TCT
+  words[0] = TXN_OPC_TCT;
```

Comment:
What about using `aie-rt` to create the transactions instead of doing it manually?

Reply:
So you and I chit-chatted about this right but I actually never saw it all the way through - does [all the transaction stuff in`aie-rt`](https://github.com/Xilinx/aie-rt/blob/1ad203de0b7f282b1c0659fd2ae3f218652c7274/driver/src/global/xaiegbl.h#L579-L585) actually refer to the transaction interpreter in the FW? But that's a good reminder - I'll check it out.

Reply:
Yeah, I was wondering about the state of it. And yes, these submit transactions refer to the transaction interpreter in the FW: https://gitenterprise.xilinx.com/ai-engine/aie-ipu/blob/master/dpu/aie_controller/include/ps/interpreter_op_impl.h#L867

Reply:
I mean I knew/know that - that `XAIE_IO_CUSTOM_OP` whatever hits that path in the firmware. What I'm saying I don't know is if the stuff in `aie-rt` that goes by the same name is actually about the "transaction interpreter" in the firmware. 

It's an easy thing to check (just look in the source of `aie-rt` to see what happens eg when you call `SubmitSerializedTransaction`) I just haven't done it yet.

Reply:
> It's an easy thing to check (just look in the source of aie-rt to see what happens eg when you call SubmitSerializedTransaction) I just haven't done it yet.

Oh, I think I see what you mean. I don't think `SubmitSerializedTransaction` would work (I hope it would though). What should work is exporting the transaction with `XAie_ExportSerializedTransaction`: https://github.com/Xilinx/aie-rt/blob/1ad203de0b7f282b1c0659fd2ae3f218652c7274/driver/src/global/xaiegbl.h#L582, which should give you the instruction bytestream like you have it here: https://github.com/nod-ai/iree-amd-aie/pull/460/files#diff-bcbdafe57dbdd0686e4a2a056c5018c13dc4164bba50861a04e92b0066268b3cR490

Reply:
Yea so this is actually exactly why I never finished this same sprint in `mlir-aie` and actually *exactly what gave me the ideas that I'm implementing here today*. I.e., if we remove everything below objectfifo then we only have `aie-rt` APIs and we can just call them and "record" them using `SerializedTransaction` and then send them as NPU/IPU instructions. The problem of course is what's mentioned up-thread here (not everything can be a runtime instruction) and CDO object and `SerializedTransaction` aren't 1-1 ([cdo has its own stupid op code suffixes](https://github.com/Xilinx/bootgen/blob/master/cdo-driver/cdo_driver.h#L27-L31)) so converting between isn't completely trivial.

Reply:
> The problem of course is what's mentioned up-thread here (not everything can be a runtime instruction) and CDO object and SerializedTransaction aren't 1-1 ([cdo has its own stupid op code prefixes](https://github.com/Xilinx/bootgen/blob/master/cdo-driver/cdo_driver.h#L27-L31)) so converting between isn't completely trivial.

I am not following this. Why do we need to translate between CDO and transactions? The serialized transaction should just be usable with the current flow by using it to create the `sdt::vector<uint32_t> instructions` object and keeping everything else the same?

Reply:
Yes but using `SerializedTransaction` just here for these few ops is +/- no net gain really. Ideally you want to use `SerializedTransaction` for *everything*. But you can't because some things *have to* go into a CDO and some things can be either/or.

Reply:
Anyway you're right - maybe the net gain is simply starting the gradual transition to using `SerializedTransaction` in as many places as possible. Lemme create and issue for myself to do what you said so I remember.